physical/postgresql/postgresql.go

```diff
@@ -6,31 +6,76 @@ import (
 	"fmt"
 	"strconv"
 	"strings"
+	"sync"
 	"time"
 
 	"github.com/hashicorp/errwrap"
 	"github.com/hashicorp/vault/physical"
 
 	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-uuid"
 
 	metrics "github.com/armon/go-metrics"
 	"github.com/lib/pq"
 )
 
+const (
+
+	// (seconds) The lock TTL matches the default that Consul API uses, 15 seconds.
+	// used in as part of SQL commands to set/extend lock expiryt time relative to database clock
+	PostgreSQLLockTTL = 15
+
+	// The amount of time to wait between the lock renewals
+	PostgreSQLLockRenewInterval = 5 * time.Second
+
+	// PostgreSQLLockRetryInterval is the amount of time to wait
+	// if a lock fails before trying again.
+	PostgreSQLLockRetryInterval = time.Second
+	// PostgreSQLWatchRetryMax is the number of times to re-try a
+	// failed watch before signaling that leadership is lost.
+	PostgreSQLWatchRetryMax = 5
+	// PostgreSQLWatchRetryInterval is the amount of time to wait
+	// if a watch fails before trying again.
+	PostgreSQLWatchRetryInterval = 5 * time.Second
+)
+
 // Verify PostgreSQLBackend satisfies the correct interfaces
 var _ physical.Backend = (*PostgreSQLBackend)(nil)
 
+//
+// HA backend was implemented based on the DynamoDB backend pattern
+// With distinction using central postgres clock, hereby avoiding
+// possible issues with multiple clocks
+//
+var _ physical.HABackend = (*PostgreSQLBackend)(nil)
+var _ physical.Lock = (*PostgreSQLLock)(nil)
+
 // PostgreSQL Backend is a physical backend that stores data
 // within a PostgreSQL database.
 type PostgreSQLBackend struct {
-	table        string
-	client       *sql.DB
-	put_query    string
-	get_query    string
-	delete_query string
-	list_query   string
-	logger       log.Logger
-	permitPool   *physical.PermitPool
+	table                    string
+	client                   *sql.DB
+	put_query                string
+	get_query                string
+	delete_query             string
+	list_query               string
+	haGetLockIdentityQuery   string
+	haGetLockValueQuery      string
+	haUpsertLockIdentityExec string
+	haDeleteLockExec         string
+
+	haEnabled  bool
+	logger     log.Logger
+	permitPool *physical.PermitPool
+}
+
+// PostgreSQLLock implements a lock using an PostgreSQL client.
+type PostgreSQLLock struct {
+	backend    *PostgreSQLBackend
+	value, key string
+	identity   string
+	held       bool
```

Comment:
I worry that the `held` field is going to be a source of issues, wherein we think we have the lock but PG disagrees.  Can we eliminate it?  It feels like a premature performance optimization.

Reply:
Yes, totally agree. Remove it.